apps/server/src/modules/user-migration/service/user-migration.service.ts

```diff
@@ -113,9 +129,41 @@ export class UserMigrationService {
 		return combinedUri;
 	}
 
-	private getMigrationRedirectUri(systemId: string): string {
-		const combinedUri = new URL(this.hostUrl);
-		combinedUri.pathname = `/api/v3/sso/oauth/${systemId}/migration`;
+	getMigrationRedirectUri(systemId: string): string {
+		const combinedUri = new URL(this.apiUrl);
+		combinedUri.pathname = `api/v3/sso/oauth/${systemId}/migration`;
 		return combinedUri.toString();
 	}
+
+	async migrateUser(currentUserId: string, externalUserId: string, targetSystemId: string): Promise<UserMigrationDto> {
+		const userDO: UserDO = await this.userService.findById(currentUserId);
+		const account: AccountDto = await this.accountService.findByUserIdOrFail(currentUserId);
+		const userDOCopy: UserDO = { ...userDO };
+		const accountCopy: AccountDto = { ...account };
+
+		try {
+			userDO.previousExternalId = userDO.externalId;
+			userDO.externalId = externalUserId;
+			userDO.lastLoginSystemChange = new Date();
+			await this.userService.save(userDO);
+			account.systemId = targetSystemId;
+			await this.accountService.save(account);
+
+			const userMigrationDto: UserMigrationDto = new UserMigrationDto({
+				redirect: `${this.hostUrl}/migration/succeed`,
+			});
+			return userMigrationDto;
+		} catch (e) {
+			await this.userService.save(userDOCopy);
+			await this.accountService.save(accountCopy);
+
+			this.logger.log(`This error occurred during migration of User: ${currentUserId} `);
+			this.logger.log(e);
+
+			const userMigrationDto: UserMigrationDto = new UserMigrationDto({
+				redirect: `${this.hostUrl}/dashboard`,
```

Comment:
see comment above.. where is this url coming from? how do we know the user is using a client that has a dashboard url, and that url happens to be a page intended for a usecase like this? Is the dashboard even reachable after a failed migration, or is the user being logged out? how do we even know the client is on the same hostUrl as the server?

These are all questions that shouldnt even concern the server...

Reply:
In our case, the user should be logged out and redirected to an errorpage, which does not exist yet. 
The whole error-handling will be prepared later, because we need more logic to redirect after logout. For this step we want just to test functionality of the migration. The dashboard url will be replaced with the errorpage in the next tickets.

Reply:
The errorpage you want to redirect to still wont be a frontend page, wont it?
So that doesnt change the point, the server cant know about the client, and it cant make assumptions about what client the user is using.

You can mention server urls in your responses as much as you want, you can pass on urls that external systems have defined or the user has saved to database, but the server can not define client redirects thtat the client is responsible for.